### PR TITLE
Feature/datanode autodiscovery

### PR DESCRIPTION
**Pull request by todvora**
This PR enables autodiscovery of datanodes that registered itself via the node-service heartbeat.

Additionally, the PR adapts the preflight/standard boot decision based on following flowchart: 

image

/jenkins-pr-deps Graylog2/graylog-plugin-enterprise#5101

## Motivation and Context
Graylog server needs to be able to detect indexer hosts (datanodes) configured during the preflight setup. These hosts  are not configured via the `elasticsearch_hosts` config property but rather loosely coupled via registration and heartbeats. 

The configuration now allows empty elasticsearch_hosts` and doesn't provide the localhost:9200 default. The default is, when needed, resolved later in the IndexerDiscoveryProvider. 

## How Has This Been Tested?
Added unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/bindings/ElasticsearchModule.java [hunk 1]
```diff
@@ -17,12 +17,18 @@
...
+import org.graylog2.configuration.IndexerHosts;
 import org.graylog2.indexer.FailureIndexMappingFactory;
 import org.graylog2.indexer.IndexMappingFactory;
 import org.graylog2.storage.DetectedSearchVersion;
-import org.graylog2.storage.providers.ElasticsearchVersionProvider;
 import org.graylog2.storage.SearchVersion;
+import org.graylog2.storage.providers.ElasticsearchVersionProvider;
+
+import java.net.URI;
+import java.util.List;
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bindings/ElasticsearchModule.java:31]: Am I right? Are those imports not used inside this class?

File: graylog2-server/src/main/java/org/graylog2/configuration/IndexerHosts.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import com.google.inject.BindingAnnotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.PARAMETER, ElementType.TYPE })
+@BindingAnnotation
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/configuration/IndexerHosts.java:28]: I am not a Guice expert, but from what I can see [here](https://github.com/google/guice/wiki/BindingAnnotations) :

```Older code may still use Guice `BindingAnnotation` in place of the standard
`@Qualifier` annotation. New code should use `@Qualifier` instead.```

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PreflightConfig.java [hunk 3]
```diff
@@ -17,15 +17,16 @@
...
 import org.graylog2.database.PersistedImpl;
 import org.graylog2.plugin.database.validators.Validator;
 
 import javax.annotation.Nullable;
 import java.util.Collections;
 import java.util.Map;
+import java.util.Optional;
 
-@CollectionName("preflight")
+@DbEntity(collection = "preflight")
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PreflightConfig.java:29]: `PreflightConfig` documents do not seem to have `title` field.
You should probably override default `titleField` property of `DBEntity` annotation.

`PreflightConfig` does not seem to be hidden behind any permissions.
Maybe you should consider changing default `readPermission` property from `NOBODY_ALOWED` to `ALL_ALOWED`?

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PreflightConfig.java [hunk 4]
```diff
@@ -36,6 +37,12 @@ protected PreflightConfig(ObjectId id, @Nullable Map<String, Object> fields) {
         super(id, fields);
     }
 
+    public PreflightConfigResult result() {
+        return Optional.ofNullable(getFields().get("result"))
+                .map(v -> PreflightConfigResult.valueOf((String) v))
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/PreflightConfig.java:42]: What about checking if "result" is really a String, to be on the safe side?

File: graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java [hunk 5]
```diff
@@ -180,9 +181,9 @@ private void runPreflightWeb(List<Module> preflightCheckModules) {
     }
 
     private void doRunWithPreflightInjector(Injector preflightInjector) {
-        final PreflightConfigService preflightConfigService = preflightInjector.getInstance(PreflightConfigService.class);
+        final PreflightBoot preflightBoot = preflightInjector.getInstance(PreflightBoot.class);
 
-        final Supplier<Boolean> shouldRun = () -> preflightConfigService.getPersistedConfig().isEmpty() || configuration.enablePreflightWebserver();
+        final Supplier<Boolean> shouldRun = preflightBoot::shouldRunPreflightWeb;
 
         if (!shouldRun.get()) {
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/ServerBootstrap.java:188]: I believe you do not need `Supplier` now, you can just call `shouldRunPreflightWeb` in the if-clause.

File: graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import javax.inject.Inject;
+import javax.inject.Named;
+import java.net.URI;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+
+public class IndexerDiscoveryProvider implements Provider<List<URI>> {
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java:36]: I would vote for some unit tests for this class, with some nice mocks! :)

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightStatusResource.java [hunk 7]
```diff
@@ -37,11 +37,11 @@
 public class PreflightStatusResource {
 
     private final Version version = Version.CURRENT_CLASSPATH;
-    private final PreflightConfigService preflightConfigService;
+    private final PreflightConfigService preflightConfigServiceIf;
```
Comment (luk-kaminski) [inline graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightStatusResource.java:40]: What does that "If" that ends the name mean?

File: graylog2-server/src/test/java/org/graylog2/configuration/IndexerDiscoveryProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                .map(address -> {
+                    final Node node = Mockito.mock(Node.class);
+                    Mockito.when(node.getTransportAddress()).thenReturn(address);
+                    return node;
+                })
+                .collect(Collectors.toList());
+    }
+
+    private Supplier<Optional<PreflightConfig>> preflightConfig(@Nullable PreflightConfigResult result) {
+        return () -> Optional.ofNullable(result).map(r -> () -> r);
```
Comment (luk-kaminski) [inline graylog2-server/src/test/java/org/graylog2/configuration/IndexerDiscoveryProviderTest.java:105]: I just wonder if this magic part `r -> () -> r` is necessary?

File: graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Inject
+    public IndexerDiscoveryProvider(
+            @Named("elasticsearch_hosts") List<URI> hosts,
+            PreflightConfigService preflightConfigServiceIf,
+            NodeService nodeService) {
+        this(hosts, preflightConfigServiceIf::getPersistedConfig, () -> nodeService.allActive(Node.Type.DATANODE).values());
+    }
+
+    IndexerDiscoveryProvider(List<URI> hosts, Supplier<Optional<PreflightConfig>> preflightConfigSupplier, Supplier<Collection<Node>> activeNodesSupplier) {
```
[the remark this answers is not in the document]
Comment (luk-kaminski) [reply graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java:60]: Probably the same with NodeService.

File: graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightStatusResource.java (hunk 7 above)
Comment (todvora) [reply graylog2-server/src/main/java/org/graylog2/bootstrap/preflight/web/resources/PreflightStatusResource.java:40]: ah, refactoring related, will fix that. Thx!

File: graylog2-server/src/test/java/org/graylog2/configuration/IndexerDiscoveryProviderTest.java (hunk 8 above)
Comment (todvora) [reply graylog2-server/src/test/java/org/graylog2/configuration/IndexerDiscoveryProviderTest.java]: simplified that part so it's more readable.  The `r -> () -> r` was just mapping a value to an interface, but I agree that it looks tricky.

File: graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (todvora) [reply graylog2-server/src/main/java/org/graylog2/configuration/IndexerDiscoveryProvider.java]: ok, I can live with PreflightConfigService but don't want to support the full NodeService, mock it and deal with it in the code. So 50:50 simplification there :-)

**Review by luk-kaminski — APPROVED** [review]

**Pull request merged**